File: 442-FindAllDuplicatesinanArray.java
/*

442. Find All Duplicates in an Array
Medium

Given an array of integers, 1 ≤ a[i] ≤ n (n = size of array), some elements appear twice and others appear once.

Find all the elements that appear twice in this array.

Could you do it without extra space and in O(n) runtime?

Example:

Input:
[4,3,2,7,8,2,3,1]

Output:
[2,3]

Link : https://leetcode.com/problems/find-all-duplicates-in-an-array/submissions/

*/
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> resList = new ArrayList<>();
        for (int i =0; i<nums.length; i++){
            int index = Math.abs(nums[i]) - 1;
            if (nums[index] < 0)
                resList.add(Math.abs(index+1));
            nums[index] = -nums[index];
        }
        return resList;
    }
}

File: 442-FindAllDuplicatesinanArray/442-FindAllDuplicatesinanArray.py
"""

442. Find All Duplicates in an Array

Given an integer array nums of length n where all the integers of nums are in the range [1, n] and each integer appears at most twice, return an array of all the integers that appears twice.

You must write an algorithm that runs in O(n) time and uses only constant auxiliary space, excluding the space needed to store the output

Example 1:

Input: nums = [4,3,2,7,8,2,3,1]
Output: [2,3]

Example 2:

Input: nums = [1,1,2]
Output: [1]

Example 3:

Input: nums = [1]
Output: []

Constraints:

    n == nums.length
    1 <= n <= 105
    1 <= nums[i] <= n
    Each element in nums appears once or twice.

"""

class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        res = []
        for i, n in enumerate(nums):
            idx = abs(nums[i]) - 1
            if nums[idx] < 0:
                res.append(idx + 1)
            nums[idx] = -nums[idx]
        return res
